src/Entity/Form/SubscriptionForm.php

```diff
@@ -156,13 +211,98 @@ public function save(array $form, FormStateInterface $form_state) {
       }
     }
     catch (\Exception $e) {
-      if (($client_error = $e->getPrevious())
-        && $client_error instanceof ClientErrorException
+      $client_error = $e->getPrevious();
+
+      if ($client_error instanceof ClientErrorException
         && $client_error->getEdgeErrorCode() === 'mint.developerHasFollowingOverlapRatePlans'
       ) {
         $form_state->set('planConflicts', $this->getOverlappingProducts($e->getMessage()));
         $form_state->setRebuild(TRUE);
       }
+      // If insufficient funds error, format nicely and add link to add credit.
+      if ($client_error instanceof ClientErrorException && $client_error->getEdgeErrorCode() === static::INSUFFICIENT_FUNDS_ERROR) {
+        preg_match_all('/\[(?\'amount\'.+)\]/', $e->getMessage(), $matches);
+        $amount = $matches['amount'][0] ?? NULL;
+        $rate_plan = $this->getEntity()->getRatePlan();
+        $currency_id = $rate_plan->getCurrency()->id();
+
+        $amount_formatted = $this->monetization->formatCurrency($matches['amount'][0], $currency_id);
+        $insufficient_funds_error_message = $this->t('You have insufficient funds to purchase plan %plan. @adenndum', [
+          '%plan' => $rate_plan->label(),
+          '%amount' => $amount_formatted,
+          '@adenndum' => $amount ? $this->t('To purchase this plan you are required to add at least %amount to your account.', ['%amount' => $amount_formatted]) : '',
+        ]);
+        $this->module_handler->alter('apigee_m10n_insufficient_balance_error_message', $insufficient_funds_error_message, $this->getEntity());
```

Comment:
I am getting  an error at this line with apigee_m10n_add_credit module *not* enabled:
```
Notice: Only variables should be passed by reference in Drupal\apigee_m10n\Entity\Form\SubscriptionForm->save() (line 235 of modules/contrib/apigee_m10n/src/Entity/Form/SubscriptionForm.php).
Drupal\apigee_m10n\Entity\Form\SubscriptionForm->save(Array, Object)
```

```suggestion
        $subscription_entity = $this->getEntity();
        $this->module_handler->alter('apigee_m10n_insufficient_balance_error_message', $insufficient_funds_error_message, $subscription_entity);
```

Reply:
Fixed via git.